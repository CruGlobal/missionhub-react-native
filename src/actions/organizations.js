import {
  GET_ORGANIZATIONS_CONTACTS_REPORT,
  GET_ORGANIZATION_MEMBERS,
  DEFAULT_PAGE_LIMIT,
  LOAD_ORGANIZATIONS,
  ORGANIZATION_CONTACTS_SEARCH,
  SURVEY_CONTACTS_SEARCH,
  RESET_ORGANIZATION_CONTACTS,
  RESET_SURVEY_CONTACTS,
} from '../constants';
import { timeFilter } from '../utils/filters';
import { organizationSelector } from '../selectors/organizations';

import callApi, { REQUESTS } from './api';

const getOrganizationsQuery = {
  limit: 100,
  include: '',
  filters: {
    descendants: false,
  },
};

export function getMyCommunities() {
  return async dispatch => {
    await dispatch(getMyOrganizations());
    return dispatch(getOrganizationsContactReports());
  };
}

export function getMyOrganizations() {
  return async (dispatch, getState) => {
    const { response: orgs } = await dispatch(
      callApi(REQUESTS.GET_ORGANIZATIONS, getOrganizationsQuery),
    );
    const orgOrder = getState().auth.person.user.organization_order;

    if (orgOrder) {
      orgs.sort((a, b) => {
        const aIndex = orgOrder.indexOf(a.id);
        const bIndex = orgOrder.indexOf(b.id);

        if (aIndex === -1) {
          return bIndex === -1 ? 0 : 1;
        }

        if (bIndex === -1) {
          return aIndex === -1 ? 0 : -1;
        }

        return aIndex > bIndex ? 1 : -1;
      });
    }

    dispatch({
      type: LOAD_ORGANIZATIONS,
      orgs,
    });
  };
}

export function getOrganizationsContactReports() {
  return async dispatch => {
    const { response } = await dispatch(
      callApi(REQUESTS.GET_ORGANIZATION_INTERACTIONS_REPORT, { period: 'P1W' }),
    );

    dispatch({
      type: GET_ORGANIZATIONS_CONTACTS_REPORT,
      reports: response.map(r => ({
        id: `${r.organization_id}`,
        contactsCount: r.contact_count,
        unassignedCount: r.unassigned_count,
        uncontactedCount: r.uncontacted_count,
      })),
    });
    return response;
  };
}

export function getOrganizationContacts(orgId, name, filters = {}) {
  const query = {
    filters: {
      permissions: 'no_permission',
      organization_ids: orgId,
    },
    include:
      'reverse_contact_assignments,reverse_contact_assignments.organization,organizational_permissions',
  };
  if (name) {
    query.filters.name = name;
  }

  const answerFilters = getAnswersFromFilters(filters);
  if (answerFilters) {
    query.filters.answer_sheets = { answers: answerFilters };
  }
  if (filters.survey) {
    query.filters.answer_sheets = {
      ...(query.filters.answer_sheets || {}),
      survey_ids: filters.survey.id,
    };

    // If there is a survey AND we're filtering by time, apply the time filter to the answer_sheets
    if (filters.time) {
      const dates = timeFilter(filters.time.value);
      query.filters.answer_sheets = {
        ...(query.filters.answer_sheets || {}),
        created_at: [dates.first, dates.last],
      };
    }
  } else {
    // TODO: Enable this when the API supports sorting contacts by `updated_at`
    //   if (filters.time) {
    //     const dates = timeFilter(filters.time.value);
    //     query.filters.updated_at = [dates.first, dates.last];
    //   }
  }
  if (filters.gender) {
    query.filters.genders = filters.gender.id;
  }
  if (filters.archived) {
    query.filters.include_archived = true;
  }
  if (filters.unassigned) {
    query.filters.assigned_tos = 'unassigned';
  }
  if (filters.uncontacted) {
    query.filters.statuses = 'uncontacted';
  }
  if (filters.labels) {
    query.filters.label_ids = filters.labels.id;
  }
  if (filters.groups) {
    query.filters.group_ids = filters.groups.id;
  }

  return async (dispatch, getState) => {
    const { organizations } = getState();
    const org = organizationSelector({ organizations }, { orgId });
    const pagination = org.contactPagination || { page: 0, hasNextPage: true };

    const offset = DEFAULT_PAGE_LIMIT * pagination.page;

    query.page = {
      limit: DEFAULT_PAGE_LIMIT,
      offset,
    };

    const { response, meta } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_LIST, query),
    );

    return dispatch({
      type: filters.survey
        ? SURVEY_CONTACTS_SEARCH
        : ORGANIZATION_CONTACTS_SEARCH,
      orgId,
      contacts: response,
      query,
      meta,
    });
  };
}

export function reloadOrganizationContacts(orgId, name, filters = {}) {
  return (dispatch, getState) => {
    const { organizations } = getState();
    const org = organizationSelector({ organizations }, { orgId });

    if (filters.survey && filters.survey.id) {
      const surveyId = filters.survey.id;
      if (
        org.surveys &&
        org.surveys[surveyId] &&
        org.surveys[surveyId].contactPagination
      ) {
        dispatch(resetContactPagination(orgId, surveyId));
      }
    }

    if (org && org.contactPagination) {
      dispatch(resetContactPagination(orgId));
    }
    return dispatch(getOrganizationContacts(orgId, name, filters));
  };
}

const resetContactPagination = (orgId, surveyId = undefined) => {
  return {
    type: surveyId ? RESET_SURVEY_CONTACTS : RESET_ORGANIZATION_CONTACTS,
    orgId,
    surveyId,
  };
};

//each question/answer filter must be in the URL in the form:
//filters[answers][questionId][]=answerTexts
function getAnswersFromFilters(filters) {
  const arrFilters = Object.keys(filters).map(k => filters[k]);
  const answers = arrFilters.filter(f => f.isAnswer);
  if (answers.length === 0) {
    return null;
  }
  let answerFilters = {};
  answers.forEach(f => {
    answerFilters[f.id] = [f.text];
  });
  return answerFilters;
}

//todo probably should start storing this stuff in Redux
export function getOrganizationMembers(orgId, query = {}) {
  const newQuery = {
    ...query,
    filters: {
      permissions: 'admin,user',
      organization_ids: orgId,
    },
    include: 'organizational_permissions',
  };
  return async dispatch => {
    const { response: members, meta } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_LIST, newQuery),
    );

    const memberIds = members.map(m => m.id);
    const reportQuery = {
      people_ids: memberIds.join(','),
      period: 'P1Y',
      organization_ids: orgId,
    };
    const { response: reportResponse } = await dispatch(
      callApi(REQUESTS.GET_PEOPLE_INTERACTIONS_REPORT, reportQuery),
    );

    // Get an object with { [key = person_id]: [value = { counts }] }
    const reportsCountObj = reportResponse.reduce(
      (p, n) => ({
        ...p,
        [`${n.person_id}`]: {
          contact_count: n.contact_count,
          uncontacted_count: n.uncontacted_count,
          contacts_with_interaction_count: n.contacts_with_interaction_count,
        },
      }),
      {},
    );
    // Merge the counts into the members array
    const membersWithCounts = members.map(m => ({
      ...m,
      ...(reportsCountObj[m.id] || {}),
    }));

    dispatch({
      type: GET_ORGANIZATION_MEMBERS,
      orgId,
      members: membersWithCounts,
      query: newQuery,
      meta,
    });
    return membersWithCounts;
  };
}

export function getOrganizationMembersNextPage(orgId) {
  return (dispatch, getState) => {
    const { page, hasNextPage } = getState().organizations.membersPagination;
    if (!hasNextPage) {
      // Does not have more data
      return Promise.resolve();
    }
    const query = {
      page: {
        limit: DEFAULT_PAGE_LIMIT,
        offset: DEFAULT_PAGE_LIMIT * page,
      },
    };
    return dispatch(getOrganizationMembers(orgId, query));
  };
}

export function addNewContact(data) {
  return (dispatch, getState) => {
    const {
      person: { id: myId },
    } = getState().auth;
    if (!data || !data.firstName) {
      return Promise.reject(
        `Invalid Data from addNewContact: no data or no firstName passed in`,
      );
    }
    let included = [];
    included.push({
      type: 'contact_assignment',
      attributes: {
        assigned_to_id: myId,
        organization_id: data.orgId || undefined,
      },
    });
    if (data.orgId) {
      included.push({
        type: 'organizational_permission',
        attributes: {
          organization_id: data.orgId,
          permission_id: data.orgPermission && data.orgPermission.permission_id,
        },
      });
    }
    if (data.email) {
      included.push({
        type: 'email',
        attributes: { email: data.email },
      });
    }
    if (data.phone) {
      included.push({
        type: 'phone_number',
        attributes: {
          number: data.phone,
          location: 'mobile',
        },
      });
    }
    const bodyData = {
      data: {
        type: 'person',
        attributes: {
          first_name: data.firstName,
          last_name: data.lastName || undefined,
          gender: data.gender || undefined,
        },
      },
      included,
    };
    const query = {};
    return dispatch(callApi(REQUESTS.ADD_NEW_PERSON, query, bodyData));
  };
}
