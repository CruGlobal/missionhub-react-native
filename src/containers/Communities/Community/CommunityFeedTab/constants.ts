export const COMMUNITY_FEED = 'nav/COMMUNITY_FEED';
