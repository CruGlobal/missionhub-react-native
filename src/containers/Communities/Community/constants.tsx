import React from 'react';
import i18next from 'i18next';

import GroupChallenges, {
  COMMUNITY_CHALLENGES,
} from '../../Groups/GroupChallenges';

import {
  CommunityFeedTab,
  COMMUNITY_FEED,
} from './CommunityFeedTab/CommunityFeedTab';
import {
  COMMUNITY_IMPACT,
  CommunityImpactTab,
} from './CommunityImpactTab/CommunityImpactTab';
import { CommunitiesCollapsibleHeaderContext } from './CommunityHeader/CommunityHeader';

// Moved here to prevent circular dependency
export const COMMUNITY_TABS = 'nav/COMMUNITY_TABS';

export const communityTabs = [
  {
    name: i18next.t('communityTabs:feed'),
    navigationAction: COMMUNITY_FEED,
    component: () => (
      <CommunityFeedTab
        collapsibleHeaderContext={CommunitiesCollapsibleHeaderContext}
      />
    ),
  },
  {
    name: i18next.t('communityTabs:challenges'),
    navigationAction: COMMUNITY_CHALLENGES,
    component: GroupChallenges,
  },
  {
    name: i18next.t('communityTabs:impact'),
    navigationAction: COMMUNITY_IMPACT,
    component: () => (
      <CommunityImpactTab
        collapsibleHeaderContext={CommunitiesCollapsibleHeaderContext}
      />
    ),
  },
];
