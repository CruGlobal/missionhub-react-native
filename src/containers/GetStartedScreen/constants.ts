export const GET_STARTED_SCREEN = 'nav/GET_STARTED';
