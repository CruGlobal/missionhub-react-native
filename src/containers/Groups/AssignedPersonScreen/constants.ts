export const CONTACT_PERSON_SCREEN = 'nav/CONTACT_PERSON';
export const IS_USER_CREATED_MEMBER_PERSON_SCREEN =
  'nav/IS_USER_CREATED_MEMBER_PERSON';
export const IS_GROUPS_MEMBER_PERSON_SCREEN = 'nav/IS_GROUPS_MEMBER_PERSON';
export const MEMBER_PERSON_SCREEN = 'nav/MEMBER_PERSON';
export const ME_PERSONAL_PERSON_SCREEN = 'nav/ME_PERSONAL_PERSON';
export const IS_GROUPS_ME_COMMUNITY_PERSON_SCREEN =
  'nav/IS_GROUPS_ME_COMMUNITY_PERSON';
export const ME_COMMUNITY_PERSON_SCREEN = 'nav/ME_COMMUNITY_PERSON';
