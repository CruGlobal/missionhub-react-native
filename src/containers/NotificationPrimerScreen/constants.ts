export const NOTIFICATION_PRIMER_SCREEN = 'nav/NOTIFICATION_PRIMER';
