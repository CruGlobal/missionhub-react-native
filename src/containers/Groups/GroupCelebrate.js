import React, { Component } from 'react';
import { SafeAreaView } from 'react-native';
import { connect } from 'react-redux';
import { translate } from 'react-i18next';
import moment from 'moment';

import CelebrateFeed from '../../components/CelebrateFeed';
import {
  getGroupCelebrateFeed,
  reloadGroupCelebrateFeed,
} from '../../actions/celebration';
import { refreshCommunity } from '../../actions/organizations';
import { organizationSelector } from '../../selectors/organizations';
import { celebrationSelector } from '../../selectors/celebration';
import { momentUtc, refresh } from '../../utils/common';
import { GLOBAL_COMMUNITY_ID } from '../../constants';
import { getReportedComments } from '../../actions/celebrateComments';

@translate('groupsCelebrate')
export class GroupCelebrate extends Component {
  state = { refreshing: false };

  componentDidMount() {
    if (this.shouldLoadFeed()) {
      this.loadItems();
    }
  }

  shouldLoadFeed = () => {
    const { pagination, celebrateItems } = this.props;

    return (
      !celebrateItems ||
      celebrateItems.length === 0 ||
      pagination.page === 0 ||
      moment().diff(momentUtc(celebrateItems[0].date), 'days', true) > 1
    );
  };

  loadItems = () => {
    const { dispatch, organization } = this.props;
    dispatch(getGroupCelebrateFeed(organization.id));
  };

  reloadItems = () => {
    const { dispatch, organization } = this.props;
    dispatch(refreshCommunity(organization.id));
    dispatch(getReportedComments(organization.id));
    return dispatch(reloadGroupCelebrateFeed(organization.id));
  };

  refreshItems = () => {
    refresh(this, this.reloadItems);
  };

  render() {
    const { refreshing } = this.state;
    const { celebrateItems, organization } = this.props;

    return (
      <SafeAreaView style={{ flex: 1 }}>
        <CelebrateFeed
          organization={organization}
          items={celebrateItems}
          loadMoreItemsCallback={this.loadItems}
          refreshCallback={this.refreshItems}
          refreshing={refreshing}
          itemNamePressable={organization.id !== GLOBAL_COMMUNITY_ID}
        />
      </SafeAreaView>
    );
  }
}

export const mapStateToProps = ({ organizations }, { organization = {} }) => {
  const selectorOrg =
    organizationSelector({ organizations }, { orgId: organization.id }) ||
    organization;

  const celebrateItems = celebrationSelector({
    celebrateItems: selectorOrg.celebrateItems || [],
  });

  return {
    celebrateItems,
    pagination: selectorOrg && selectorOrg.celebratePagination,
  };
};

export default connect(mapStateToProps)(GroupCelebrate);
