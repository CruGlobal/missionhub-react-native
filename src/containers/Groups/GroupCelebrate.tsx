import React from 'react';
import { connect, useDispatch } from 'react-redux';

import CelebrateFeed from '../CelebrateFeed';
import { TrackStateContext } from '../../actions/analytics';
import { refreshCommunity } from '../../actions/organizations';
import { organizationSelector } from '../../selectors/organizations';
import {
  orgIsGlobal,
  shouldQueryReportedComments,
  getAnalyticsPermissionType,
} from '../../utils/common';
import { ANALYTICS_PERMISSION_TYPE } from '../../constants';
import { getReportedComments } from '../../actions/reportComments';
import { orgPermissionSelector } from '../../selectors/people';
import { AuthState } from '../../reducers/auth';
import { Organization, OrganizationsState } from '../../reducers/organizations';
import { useAnalytics } from '../../utils/hooks/useAnalytics';

export interface GroupCelebrateProps {
  organization: Organization;
  shouldQueryReport: boolean;
  analyticsPermissionType: TrackStateContext[typeof ANALYTICS_PERMISSION_TYPE];
}

const GroupCelebrate = ({
  organization,
  shouldQueryReport,
  analyticsPermissionType,
}: GroupCelebrateProps) => {
  useAnalytics({
    screenName: ['community', 'celebrate'],
    screenContext: { [ANALYTICS_PERMISSION_TYPE]: analyticsPermissionType },
  });
  const dispatch = useDispatch();

  const handleRefetch = () => {
    dispatch(refreshCommunity(organization.id));
    shouldQueryReport && dispatch(getReportedComments(organization.id));
  };

  return (
    <CelebrateFeed
      testID="CelebrateFeed"
      organization={organization}
      onRefetch={handleRefetch}
      itemNamePressable={!orgIsGlobal(organization)}
    />
  );
};

const mapStateToProps = (
  {
    auth,
    organizations,
  }: { auth: AuthState; organizations: OrganizationsState },
  { orgId }: { orgId: string },
) => {
  const organization = organizationSelector({ organizations }, { orgId });
  const myOrgPermission = orgPermissionSelector(
    {},
    { person: auth.person, organization },
  );

  return {
    organization,
    shouldQueryReport: shouldQueryReportedComments(
      organization,
      myOrgPermission,
    ),
    analyticsPermissionType: getAnalyticsPermissionType(myOrgPermission),
  };
};

export default connect(mapStateToProps)(GroupCelebrate);
