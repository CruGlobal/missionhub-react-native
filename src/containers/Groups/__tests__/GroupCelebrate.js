import React from 'react';
import MockDate from 'mockdate';

import GroupCelebrate from '../GroupCelebrate';
import {
  renderShallow,
  testSnapshotShallow,
  createThunkStore,
} from '../../../../testUtils';
import {
  reloadGroupCelebrateFeed,
  getGroupCelebrateFeed,
} from '../../../actions/celebration';
import { getReportedComments } from '../../../actions/reportComments';
import { refreshCommunity } from '../../../actions/organizations';
import * as common from '../../../utils/common';
import { INTERACTION_TYPES, CELEBRATEABLE_TYPES } from '../../../constants';

jest.mock('../../../actions/organizations');
jest.mock('../../../actions/celebration');
jest.mock('../../../actions/reportComments');

MockDate.set('2017-06-18');
const celebrate1 = {
  id: '10',
  celebrateable_type: CELEBRATEABLE_TYPES.completedInteraction,
  changed_attribute_value: '2018-06-11 00:00:00 UTC',
  adjective_attribute_value:
    INTERACTION_TYPES.MHInteractionTypeDiscipleshipConversation.id,
};
const celebrate2 = {
  id: '20',
  celebrateable_type: CELEBRATEABLE_TYPES.completedStep,
  changed_attribute_value: '2018-06-10 00:00:00 UTC',
};
const celebrate3 = {
  id: '30',
  celebrateable_type: CELEBRATEABLE_TYPES.acceptedCommunityChallenge,
  changed_attribute_value: '2018-06-10 00:00:00 UTC',
};

const celebratePagination = {
  hasNextPage: true,
  page: 1,
};

const orgId = '1';
const org = {
  id: orgId,
  celebrateItems: [celebrate1, celebrate2, celebrate3],
  celebratePagination: celebratePagination,
};

const store = { organizations: { all: [org] } };

getReportedComments.mockReturnValue(() => ({ type: 'got repoerted comments' }));
reloadGroupCelebrateFeed.mockReturnValue(() => Promise.resolve());
getGroupCelebrateFeed.mockReturnValue({ type: 'got group celebrate feed' });
refreshCommunity.mockReturnValue({ type: 'refreshed community' });

it('should render correctly', () => {
  testSnapshotShallow(
    <GroupCelebrate orgId={orgId} store={createThunkStore(store)} />,
  );
});

it('should render empty correctly', () => {
  testSnapshotShallow(
    <GroupCelebrate
      orgId={orgId}
      store={createThunkStore({
        ...store,
        organizations: { all: [{ ...org, celebrateItems: [] }] },
      })}
    />,
  );
});

it('should refresh correctly', async () => {
  const component = renderShallow(
    <GroupCelebrate orgId={orgId} store={createThunkStore(store)} />,
    store,
  );

  await component
    .childAt(1)
    .props()
    .refreshCallback();

  expect(refreshCommunity).toHaveBeenCalledWith(org.id);
  expect(getReportedComments).toHaveBeenCalledWith(org.id);
  expect(reloadGroupCelebrateFeed).toHaveBeenCalledWith(org.id);
});

it('should refresh items properly', () => {
  const instance = renderShallow(
    <GroupCelebrate orgId={orgId} store={createThunkStore(store)} />,
    store,
  ).instance();

  common.refresh = jest.fn();
  instance.refreshItems();

  expect(common.refresh).toHaveBeenCalledWith(instance, instance.reloadItems);
});
