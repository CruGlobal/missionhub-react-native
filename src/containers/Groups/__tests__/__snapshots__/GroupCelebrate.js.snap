// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<CelebrateFeed
  items={
    Array [
      Object {
        "date": "2018-06-11 00:00:00 UTC",
        "events": Array [
          Object {
            "celebrateable_type": "interaction",
            "changed_attribute_value": "2018-06-11 00:00:00 UTC",
            "id": "10",
          },
        ],
        "id": "1",
      },
      Object {
        "date": "2018-06-10 00:00:00 UTC",
        "events": Array [
          Object {
            "celebrateable_type": "accepted_challenge",
            "changed_attribute_value": "2018-06-10 00:00:00 UTC",
            "id": "20",
          },
          Object {
            "celebrateable_type": "contact_assignment",
            "changed_attribute_value": "2018-06-10 00:00:00 UTC",
            "id": "30",
          },
        ],
        "id": "2",
      },
    ]
  }
  loadMoreItemsCallback={[Function]}
  refreshCallback={[Function]}
/>
`;
